minion_data/preparation/_old_metrichon.py

```diff
@@ -0,0 +1,221 @@
+from glob import glob
+from minion_data.preparation import bioinf_utils
+from pprint import pprint
+import numpy as np
+import h5py
+from typing import *
+from minion_data import dataset_pb2
+
+
+def find_ref(fast5_path):
+    dirname = os.path.dirname(fast5_path)
+    basename = os.path.basename(fast5_path)
+    name, ext = os.path.splitext(basename)
+    ref_path = os.path.join(dirname, name + '.ref')
+
+    if not os.path.exists(ref_path):
+        raise ValueError(
+            "{}-> search query:{}.ref in following directory {}".format(basename, name, dirname))
+    return ref_path
+
+
+def get_basecalled_data(h5, target_read, sampling_rate):
+    basecalled_events = h5['/Analyses/Basecall_1D_000/BaseCalled_template/Events']
+    basecalled_events = np.array(basecalled_events.value[['start', 'length', 'model_state', 'move']])
+
+    raw_start_time = h5['Raw/Reads/' + target_read].attrs['start_time']
+    events = h5['/Analyses/EventDetection_000/Reads/' + target_read]
+    event_start_time = events.attrs['start_time']
+
+    basecalled_events['start'] -= events['Events'][0]['start'] / sampling_rate
+    basecalled_events['start'] += (event_start_time - raw_start_time) / sampling_rate
+    return (
+        basecalled_events,
+        h5['/Analyses/Basecall_1D_000/BaseCalled_template/Fastq'][()].decode().split('\n')
+    )
+
+
+def read_fast5(fast5_path):
+    with h5py.File(fast5_path, 'r') as h5:
+        reads = h5['Raw/Reads']
+        target_read = list(reads.keys())[0]
+        sampling_rate = h5['UniqueGlobalKey/channel_id'].attrs['sampling_rate']
+        basecalled, fastq = get_basecalled_data(h5, target_read, sampling_rate)
+
+        signal = h5['Raw/Reads/' + target_read]['Signal']
+        start_time = basecalled[0]['start']
+        start_pad = int(sampling_rate * basecalled[0]['start'])
+        signal_len = int(sampling_rate * (basecalled[-1]['start'] + basecalled[-1]['length'] - basecalled[0]['start']))
+
+        # np.testing.assert_allclose(len(signal), start_pad + signal_len, rtol=1e-2)  # Within 1% relative tolerance, TODO check for HMM and RNN discrepancy
+
+        basecalled['start'] -= start_time
+        signal = signal[start_pad:start_pad + signal_len]
+
+        return {
+            'signal': signal,
+            'basecalled': basecalled,
+            'sampling_rate': sampling_rate,
+            'fastq': fastq,
+            'start_pad': start_pad,
+            'signal_len': signal_len
+        }
+
+
+def read_fast5_raw_ref(fast5_path, ref_path=None, verify_file=True, correct=True) -> dataset_pb2.DataPoint:
+    fast5 = read_fast5(fast5_path)
+
+    signal = fast5['signal']
+    basecalled = fast5['basecalled']
+    sampling_rate = fast5['sampling_rate']
+    fastq = fast5['fastq']
+    labels: List[dataset_pb2.DataPoint.BPConfidenceInterval] = []
+
+    # Initial conditions
+    # start, length, model_state, move
+    for bp in basecalled[0]['model_state'].decode("ASCII"):  # TODO: Should this be reversed?
```

Comment:
I don't think it should be reversed - state represents basepairs present in the pore. Next state (if `move` is != 0) will have `move` bases removed from the beginning of the state string and new bases added to the back

Example:
https://github.com/mratkovic/masters-thesis/blob/master/tex/imgs/basecall_events.png?raw=true

Reply:
ok, it wasn't reversed in the minion-basecaller, so I'm just double checking here.